### ISSUE TITLE
Optimisation rule for replacing div/mul with shr/shl

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#6015]
With Constantinople we have SHL/SHR opcodes and the codegen already utilises that if shift operators are being used. There is also an optimisation rule to replace shifts of constants with a constant.

However due to a missing feature (passing EVM version) to simplification rules we do not have any to transform div/mul.

The rule should be as follows:
- `div(x, y)` where y is a power of 2 can be transformed to `shr(x, log2(y))` where `log2(y)` should be precomputed of course
- same with `mul(x, y)` -> `shl(x, log2(y))`

This should help to optimise code where shifts are used in inline assembly. One good example is ENS' [solsha1](https://github.com/ensdomains/solsha1).

@ekpyron @chriseth any ideas how to introduce EVM version dependent rules into the simplificationrules?

**Comment by ekpyron** [comment]
Might it make sense to just always add these rules and always translate to shifts and only to replace all shifts back to ``div`` and ``mul`` during EVM bytecode generation (at that stage we need to know the EVM version anyways)?

Also: we probably need more rules than that for the shifts? Like associativity rules and merging sequences of shifts when possible? What about mixing mul/div and shl/shr and so on?

**Comment by axic** [comment]
>  only to replace all shifts back to `div` and `mul` during EVM bytecode generation

Well currently these rules are applied *after* code generation.

> Like associativity rules and merging sequences of shifts when possible?

That could be useful. Could also be added right now without any risk as long as the rules depend on `shl`/`shr` being present as that suggest the code was already compiled for constantinople.

> What about mixing mul/div and shl/shr and so on?

I think this also could be added without risk as long as it depends on `shl`/`shr` being present in the first place.

**Comment by ekpyron** [comment]
Yes, I haven't thought that through - but the rules are not only applied after code generation, but also in the yul simplifier before code generation...
But anyways: we could still always convert to shifts and in the case of old evm versions just have another pass after running the rules that will replace all shifts with mul/div again. This would allow us to simplify code generation for shifts (we could always emit shift instructions, if they are automatically replaced at the very end for old evm versions).

**Comment by axic** [comment]
> This would allow us to simplify code generation for shifts (we could always emit shift instructions, if they are automatically replaced at the very end for old evm versions).

I think that might be a good idea. I think the rule list should be run after applying that "final" pass though.

**Comment by ekpyron** [comment]
More generally it might be possible to always generate byte code for the most recent evm version first and optimize that and have a dedicated pass afterwards that "downgrades" the generated byte code to be compatible with older EVM versions - we'd need to think about whether that's possible, though (e.g. for things like staticcall, etc.).

**Comment by ekpyron** [comment]
> > This would allow us to simplify code generation for shifts (we could always emit shift instructions, if they are automatically replaced at the very end for old evm versions).
> 
> I think that might be a good idea. I think the rule list should be run after applying that "final" pass though.

I'm not sure - why would we want to run this pass before optimization and not after? Ideally the result would be the same and we wouldn't need to let the rules depend on the evm version.

**Comment by axic** [comment]
After you have replaced a shift with multiplication for example it might be apparent that you can combine two multiplications (`mul shl` -> `mul mul` -> `mul`).

Also keep in mind that different "evm versions" (more like hard forks) can command different gas costs for instructions and this influences certain optimisation steps (for example the constant optimiser).

**Comment by ekpyron** [comment]
The question is: might the ``mul shl``->``mul mul``->``mul`` simplification be worth it, even if we have shifts available? That probably depends on the values :-)...

**Comment by axic** [comment]
This is the exact question I was contemplating, but if I remember correctly the rule list does a gas calculation after an iteration to see if it is worth it. So potentially could do the mul rule even if shift is used 😉

**Comment by ekpyron** [comment]
In any case, in the meantime I tend to agree that we might want to pass the EVMVersion to the rule list: even if we replace all shifts before applying the rules, we still need to disable the ``mul->shl`` rule.
The question is: what about the yul simplifier? It should probably be EVM-version-agnostic (resp. bytecode agnostic in general) - but does that mean it should do ``mul->shl`` or ``shl->mul`` :-)?

**Comment by ekpyron** [comment]
> This is the exact question I was contemplating, but if I remember correctly the rule list does a gas calculation after an iteration to see if it is worth it. So potentially could do the mul rule even if shift is used wink

Does it? Check the gas benefit I mean? I'm not too familiar with the bytecode optimizer code, but I wouldn't see where it'd do that - and it'd be tricky anyways - ``shl->mul`` would always be bad in one iteration, but might be worth it after another step doing ``mul mul->mul``... the only way to always be optimal here would probably be backtracking, wouldn't it?

If we did check the gas costs for every iteration and did backtracking, then we could just set the gas costs for shl/shr to infinite where it isn't supported and that should do it :-).

**Comment by chriseth** [comment]
The gas calculation just counts instructions, not their cost, if I remember correctly.

**Comment by axic** [comment]
This can be easily implemented after #6318.

**Comment by axic** [comment]
Hm, I actually had some local code implementing this. Will push.

**Comment by hrkrshnn** [comment]
@axic Do you have a branch implementing this?

**Comment by axic** [comment]
https://github.com/ethereum/solidity/commit/35fe5cde06ebaec4abe2e57e366495d2c8f25b21, but now it can be implemented using the `feasible` feature with less rules (see https://github.com/ethereum/solidity/issues/6015#issuecomment-474672720)

**Comment by cameel** [comment]
Looks like it was implemented in 0.8.3 by #11089.

Rules for sequences of shifts were also done in #6530, while #6756 covered mixing mul/div and shifts.
